File: services/mail/internal/app/mailproviders.go
package app

import (
	"fmt"
	"net"
	"strings"

	"github.com/sgao19/erp-go/pkg/apierr"
)

// 邮件服务商预设：从员工填的地址，推出该用哪台收发服务器。
//
// **为什么这张表在服务端。** 绑定的动作是「拿着这个授权码去登录那台主机」。
// 主机名如果由调用方给，任何一个登录了的员工都能让邮件服务带着凭据去连
// 任意 host:port——内网地址也在内。前端会有一份同样代号的清单（每家怎么
// 开 IMAP、去哪儿拿授权码），但那份只管**显示**；主机名以这张表为准，代
// 号对不上就直接拒，不会静默用一个猜出来的值。
//
// 表里的值抄自各家自己的帮助页。个人邮箱和企业邮是**两套不同的主机**
// （163.com 和企业邮 qiye.163.com 不通用），所以分开列。
type MailProvider struct {
	// Code 是前后端共用的标识。前端的清单用同一串代号，对不上时服务端
	// 报错而不是猜——两边各改一半的那种漂移，宁可当场炸。
	Code string
	// Domains 是「填了这个后缀就认得出是这家」的域名。企业邮用自己的域名
	// （me@sunrise.com 可能托管在腾讯企业邮上），认不出来，所以那几家的
	// Domains 是空的——只能由员工在表单里挑。
	Domains []string
	// Domain 是这次绑定实际用的发信域名，从要绑的地址里取。表里的条目不
	// 填它——同一家服务商托管着无数个域名。
	Domain string

	SMTPHost     string
	SMTPPort     int32
	SMTPSecurity string
	IMAPHost     string
	IMAPPort     int32
	IMAPSecurity string

	// NeedsOAuth 的服务商不能用密码绑：它们已经关掉了 IMAP 的基本认证，
	// 拿密码去登只会得到一句和「密码错了」长得一模一样的拒绝。与其让员工
	// 反复试，不如在绑之前就说清楚该走哪扇门。
	NeedsOAuth bool
}

// mailProviders 是全部认得的服务商。顺序就是前端下拉框的顺序：企业邮在前
// （公司自己的箱是主要用途），个人邮箱在后。
var mailProviders = []MailProvider{
	{
		Code: "p263",
		// 263 企业邮的域名是各家公司自己的，认不出来。
		SMTPHost: "smtp.263.net", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.263.net", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:     "tencent",
		SMTPHost: "smtp.exmail.qq.com", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.exmail.qq.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:     "ali",
		SMTPHost: "smtp.qiye.aliyun.com", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.qiye.aliyun.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code: "netease",
		// 网易企业邮的 SMTP over SSL 是 994，不是常见的 465——照它自己
		// 的帮助页写。
		SMTPHost: "smtp.qiye.163.com", SMTPPort: 994, SMTPSecurity: "SSL",
		IMAPHost: "imap.qiye.163.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:    "gmail",
		Domains: []string{"gmail.com", "googlemail.com"},
		// Google 早就关了「用账号密码登 IMAP」。这里留着主机配置是因为
		// 应用专用密码（App Password）仍然走这条路；而开了两步验证才能
		// 生成应用专用密码，所以前端要把这句提示写出来。
		SMTPHost: "smtp.gmail.com", SMTPPort: 587, SMTPSecurity: "STARTTLS",
		IMAPHost: "imap.gmail.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:     "qq",
		Domains:  []string{"qq.com", "vip.qq.com", "foxmail.com"},
		SMTPHost: "smtp.qq.com", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.qq.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:     "netease163",
		Domains:  []string{"163.com"},
		SMTPHost: "smtp.163.com", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.163.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:     "netease126",
		Domains:  []string{"126.com"},
		SMTPHost: "smtp.126.com", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.126.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:     "sina",
		Domains:  []string{"sina.com", "sina.cn"},
		SMTPHost: "smtp.sina.com", SMTPPort: 465, SMTPSecurity: "SSL",
		IMAPHost: "imap.sina.com", IMAPPort: 993, IMAPSecurity: "SSL",
	},
	{
		Code:    "outlook",
		Domains: []string{"outlook.com", "hotmail.com", "live.com", "msn.com"},
		// 微软 2024 年起关掉了个人账号的 IMAP 基本认证，密码和应用专用
		// 密码**都不行**，只剩 OAuth。主机配置照留，等哪天接上微软的
		// OAuth 就能用；在那之前 NeedsOAuth 让绑定入口把话说在前面，
		// 而不是让员工对着一句「授权码错误」猜半天。
		SMTPHost: "smtp-mail.outlook.com", SMTPPort: 587, SMTPSecurity: "STARTTLS",
		IMAPHost: "outlook.office365.com", IMAPPort: 993, IMAPSecurity: "SSL",
		NeedsOAuth: true,
	},
}

// ProviderCodeOther 是「都不是上面这些」那一档：主机由员工自己填。
//
// 老板要的「包括其他的」就是它。这一档比其他档危险，因为主机名回到了调用
// 方手里——所以它单独走 validateCustomHost，只放行公网主机名和标准邮件端口。
const ProviderCodeOther = "other"

// MailProviderByCode 按代号取预设。第二个返回值是「认不认得这个代号」。
func MailProviderByCode(code string) (MailProvider, bool) {
	code = strings.TrimSpace(strings.ToLower(code))
	for _, p := range mailProviders {
		if p.Code == code {
			return p, true
		}
	}
	return MailProvider{}, false
}

// MailProviderForAddress 按地址后缀猜服务商，猜不出来返回 false。
//
// 只对个人邮箱有用。企业邮用各家自己的域名，从 me@sunrise.com 看不出它托
// 管在腾讯还是 263，那种情况必须由员工在表单里挑——猜错的代价是拿着凭据
// 去登了另一家的服务器。
func MailProviderForAddress(email string) (MailProvider, bool) {
	at := strings.LastIndex(email, "@")
	if at < 0 {
		return MailProvider{}, false
	}
	domain := strings.TrimSpace(strings.ToLower(email[at+1:]))
	if domain == "" {
		return MailProvider{}, false
	}
	for _, p := range mailProviders {
		for _, d := range p.Domains {
			if d == domain {
				return p, true
			}
		}
	}
	return MailProvider{}, false
}

// MailProviderCodes 是全部代号，给测试和前端对表用。
func MailProviderCodes() []string {
	out := make([]string, 0, len(mailProviders))
	for _, p := range mailProviders {
		out = append(out, p.Code)
	}
	return out
}

// mailPorts 是允许连的端口。收窄到这几个不是洁癖：这一档的主机名来自员工，
// 不限端口的话，「绑个邮箱」就成了让服务器去敲内网任意端口的一个按钮。
//
//	993 IMAPS · 143 IMAP+STARTTLS · 465 SMTPS · 587 SMTP+STARTTLS
//	994 网易企业邮的 SMTPS
var mailPorts = map[int32]bool{143: true, 465: true, 587: true, 993: true, 994: true}

// validateCustomHost 检查员工手填的主机。
//
// 挡两件事：
//
//   - **内网。** 直接写 IP 的一律拒（127.0.0.1、169.254.169.254 这类云元数据
//     地址、10./172.16./192.168. 的内网段都在内）。只收域名，而域名会在连接
//     时才解析——解析结果指回内网仍然可能，那一层要靠出网策略挡，这里挡的是
//     最直接的那条路。
//   - **非邮件端口。** 见 mailPorts。
func validateCustomHost(host string, port int32) error {
	host = strings.TrimSpace(host)
	if host == "" {
		return apierr.Invalid("NT_MAIL_HOST_REQUIRED", "请填写服务器地址")
	}
	if len(host) > 255 {
		return apierr.Invalid("NT_MAIL_HOST_INVALID", "服务器地址过长")
	}
	// 直接写 IP 的一律拒。云元数据地址和内网段都从这条路进来，而员工填
	// 邮件服务器本来也没有写 IP 的理由。
	if ip := net.ParseIP(host); ip != nil {
		return apierr.Invalid("NT_MAIL_HOST_INVALID",
			"服务器地址请填域名，例如 imap.example.com")
	}
	lower := strings.ToLower(host)
	if lower == "localhost" || strings.HasSuffix(lower, ".localhost") ||
		strings.HasSuffix(lower, ".local") || strings.HasSuffix(lower, ".internal") {
		return apierr.Invalid("NT_MAIL_HOST_INVALID", "不能填本机或内网地址")
	}
	// 至少要有一个点：单标签主机名（"mail"）只在内网解析得出来。
	if !strings.Contains(lower, ".") {
		return apierr.Invalid("NT_MAIL_HOST_INVALID",
			"服务器地址请填完整域名，例如 imap.example.com")
	}
	for _, r := range lower {
		if !(r >= 'a' && r <= 'z') && !(r >= '0' && r <= '9') && r != '.' && r != '-' {
			return apierr.Invalid("NT_MAIL_HOST_INVALID", "服务器地址含有非法字符")
		}
	}
	if !mailPorts[port] {
		return apierr.Invalid("NT_MAIL_PORT_INVALID",
			fmt.Sprintf("端口 %d 不是常见的邮件端口，请核对服务商的说明", port))
	}
	return nil
}

File: frontend/src/lib/mailProviders.ts
// 邮件服务商清单：员工加邮箱时挑一个，服务器不用他填。
//
// **这份清单只管显示。** 真正连哪台服务器由服务端按 code 查它自己那张表
// （services/mail/internal/app/mailproviders.go）决定——主机名如果由前端发
// 上去，任何登录了的员工都能让邮件服务带着凭据去连任意 host:port。所以这里
// 的 code 必须和服务端那张表一字不差；对不上时服务端会当场报「认不出这个
// 邮件服务商」，而不是猜一个连上去。
//
// 之所以从 MailHostDialog.vue 搬出来：那个对话框是**管理员**配「一家公司
// 一套」服务器的地方，只有 iam:role:write 看得见。现在每个员工都要能给
// 自己加信箱，而且可能是别家服务商的（263 的公司里绑 Gmail），预设必须
// 出现在员工的表单上。
//
// 放在 lib/ 而不是留在 .vue 里，是因为 .vue 组件在这个仓库里没有任何单元
// 测试手段（没装 @vue/test-utils），留在里面就是零覆盖。
export interface MailProvider {
  /** 和服务端 mailproviders.go 共用的代号。 */
  code: string
  /** 界面上显示的名字。不走 i18n 动态键——见文件末尾那段注释。 */
  label: string
  /** 一句话告诉员工去哪儿拿授权码。这是这份清单最有用的部分。 */
  hint: string
  /** 认得出来的地址后缀，只用来在员工输入地址后自动选中。 */
  domains?: readonly string[]
  /** true = 这家已经关掉了密码登录，只能走它自己的授权。 */
}

export const MAIL_PROVIDERS: readonly MailProvider[] = [
  {
    code: 'p263',
    label: '263 企业邮',
    hint: '在 263 邮箱网页版「设置 → 客户端设置」里开启 IMAP，密码就是邮箱密码',
  },
  {
    code: 'tencent',
    label: '腾讯企业邮',
    hint: '在企业邮网页版「设置 → 收发信设置」里开启 IMAP，并生成一个客户端专用密码',
  },
  {
    code: 'ali',
    label: '阿里企业邮',
    hint: '在阿里邮箱「设置 → 账户与安全」里开启 IMAP，用邮箱密码登录',
  },
  {
    code: 'netease',
    label: '网易企业邮',
    hint: '在企业邮网页版「设置 → 客户端设置」里开启 IMAP，并生成客户端授权码',
  },
  {
    code: 'gmail',
    label: 'Gmail',
    hint: '推荐直接用下面的「用 Google 登录」。要填密码的话，需要先开两步验证，再生成一个「应用专用密码」——账号密码本身登不上',
    domains: ['gmail.com', 'googlemail.com'],
  },
  {
    code: 'qq',
    label: 'QQ 邮箱',
    hint: '在 QQ 邮箱「设置 → 账户」里开启 IMAP，会给你一串授权码；填授权码，不是 QQ 密码',
    domains: ['qq.com', 'vip.qq.com', 'foxmail.com'],
  },
  {
    code: 'netease163',
    label: '163 邮箱',
    hint: '在 163 邮箱「设置 → POP3/SMTP/IMAP」里开启 IMAP，会给你一串授权码；填授权码，不是登录密码',
    domains: ['163.com'],
  },
  {
    code: 'netease126',
    label: '126 邮箱',
    hint: '在 126 邮箱「设置 → POP3/SMTP/IMAP」里开启 IMAP，会给你一串授权码；填授权码，不是登录密码',
    domains: ['126.com'],
  },
  {
    code: 'sina',
    label: '新浪邮箱',
    hint: '在新浪邮箱「设置 → 客户端 POP/IMAP/SMTP」里开启 IMAP，并生成独立密码',
    domains: ['sina.com', 'sina.cn'],
  },
] as const

/** 「都不是上面这些」：服务器由员工自己填。和服务端的 ProviderCodeOther 对应。 */
export const PROVIDER_OTHER = 'other'

/**
 * 按地址后缀猜服务商，猜不出来返回 null。
 *
 * 只对个人邮箱有用。企业邮用各家公司自己的域名，从 me@sunrise.com 看不出
 * 它托管在腾讯还是 263——那种情况留空，服务端会落回这家公司自己配的那套。
 */
export function providerForAddress(email: string): MailProvider | null {
  const at = email.lastIndexOf('@')
  if (at < 0) return null
  const domain = email.slice(at + 1).trim().toLowerCase()
  if (!domain) return null
  return MAIL_PROVIDERS.find((p) => p.domains?.includes(domain)) ?? null
}

/** 按代号取，认不得返回 null。 */
export function providerByCode(code: string): MailProvider | null {
  const want = code.trim().toLowerCase()
  return MAIL_PROVIDERS.find((p) => p.code === want) ?? null
}

// 名字和提示语直接写在这里，不走 i18n 的动态键。
//
// 原来的写法是 t(`mailbox.presetNames.${p.key}`)，而 locales.test.ts 只扫
// 字面量形式的调用，模板字符串拼出来的键扫不到——加一家服务商却漏了词条，
// 测试全绿，界面上直接显示 `mailbox.presetNames.outlook` 这串键名。这个仓库
// 真出过这种事故。服务商名本来也不翻译（「263 企业邮」在英文界面里还是那
// 几个字），所以写死比补三份词条更诚实。
